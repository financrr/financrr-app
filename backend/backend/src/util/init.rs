use std::fmt::Debug;
use std::process;

use tracing::error;

pub(crate) fn expect_or_exit<T, E: Debug>(result: Result<T, E>, message: &str) -> T {
    match result {
        Ok(value) => value,
        Err(err) => {
            error!("{}: {:?}", message, err);
            process::exit(1);
        }
    }
}
