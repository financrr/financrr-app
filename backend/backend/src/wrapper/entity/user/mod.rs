use actix_web::dev::Payload;
use actix_web::{FromRequest, HttpRequest};
use futures_util::future::LocalBoxFuture;
use sea_orm::{EntityName, EntityTrait};
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use utoipa::ToSchema;

use dto::Credentials;
use entity::prelude::User as DbUser;
use entity::user;
use entity::user::Model;

use crate::api::error::api::ApiError;
use crate::database::entity::{count, find_one, find_one_or_error, insert};
use crate::permission_impl;
use crate::util::auth::extract_bearer_token;
use crate::wrapper::entity::session::Session;
use crate::wrapper::entity::user::dto::UserRegistration;
use crate::wrapper::entity::{TableName, WrapperEntity};
use crate::wrapper::permission::{Permission, Permissions};
use crate::wrapper::types::phantom::{Identifiable, Phantom};

pub(crate) mod dto;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub(crate) struct User {
    pub(crate) id: i32,
    pub(crate) username: String,
    pub(crate) email: Option<String>,
    pub(crate) display_name: Option<String>,
    #[serde(with = "time::serde::rfc3339")]
    pub(crate) created_at: OffsetDateTime,
    pub(crate) is_admin: bool,
}

impl User {
    pub(crate) async fn exists(id: i32) -> Result<bool, ApiError> {
        Ok(count(user::Entity::find_by_id(id)).await? > 0)
    }

    pub(crate) async fn authenticate(credentials: &Credentials) -> Result<Self, ApiError> {
        let user = find_one(DbUser::find_by_username(credentials.username.as_str())).await;
        match user {
            Ok(Some(user)) => {
                if user.verify_password(credentials.password.as_bytes()).unwrap_or(false) {
                    Ok(Self::from(user))
                } else {
                    Err(ApiError::InvalidCredentials())
                }
            }
            _ => Err(ApiError::InvalidCredentials()),
        }
    }

    pub(crate) async fn register(registration: UserRegistration) -> Result<Self, ApiError> {
        match user::ActiveModel::register(
            registration.username,
            registration.email,
            registration.display_name,
            registration.password,
        ) {
            Ok(user) => {
                let model = insert(user).await?;
                let user = Self::from(model);
                user.add_permission(user.id, Permissions::all()).await?;

                Ok(user)
            }
            Err(e) => Err(ApiError::from(e)),
        }
    }
}

permission_impl!(User);

impl Identifiable for User {
    async fn find_by_id(id: i32) -> Result<Self, ApiError>
    where
        Self: Sized,
    {
        find_one_or_error(user::Entity::find_by_id(id), "User").await.map(Self::from)
    }
}

impl TableName for User {
    fn table_name() -> &'static str {
        user::Entity.table_name()
    }
}

impl WrapperEntity for User {
    fn get_id(&self) -> i32 {
        self.id
    }
}

impl FromRequest for User {
    type Error = ApiError;
    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;

    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
        let req = req.clone();
        Box::pin(async move {
            // get bearer token from request
            let token = extract_bearer_token(&req)?;
            let user_id = Session::find_user_id(token).await?;

            Self::find_by_id(user_id).await
        })
    }
}

impl FromRequest for Phantom<User> {
    type Error = ApiError;
    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;

    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
        let req = req.clone();
        Box::pin(async move {
            let token = extract_bearer_token(&req)?;
            let user_id = Session::find_user_id(token).await?;
            User::exists(user_id).await?;

            Ok(Self::new(user_id))
        })
    }
}

impl From<user::Model> for User {
    fn from(value: Model) -> Self {
        Self {
            id: value.id,
            username: value.username,
            email: value.email,
            display_name: value.display_name,
            created_at: value.created_at,
            is_admin: value.is_admin,
        }
    }
}
