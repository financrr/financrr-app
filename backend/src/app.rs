use crate::initializers::openapi::OpenApiInitializer;
use crate::initializers::services::ServicesInitializer;
use crate::utils::folder::{create_necessary_folders, STORAGE_FOLDER};
use crate::{controllers, models::_entities::users, tasks, workers::downloader::DownloadWorker};
use async_trait::async_trait;
use loco_rs::cache::Cache;
use loco_rs::storage::Storage;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::{BackgroundWorker, Queue},
    boot::{create_app, BootResult, StartMode},
    cache,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    storage,
    task::Tasks,
    Result,
};
use migration::Migrator;
use mimalloc::MiMalloc;
use sea_orm::DatabaseConnection;
use std::path::Path;

#[global_allocator]
static GLOBAL: MiMalloc = MiMalloc;

pub struct App;

#[async_trait]
impl Hooks for App {
    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA").or(option_env!("GITHUB_SHA")).unwrap_or("dev")
        )
    }

    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    async fn boot(mode: StartMode, environment: &Environment) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![Box::new(OpenApiInitializer), Box::new(ServicesInitializer)])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .prefix("/api/v1")
            .add_route(controllers::auth::routes())
            .add_route(controllers::session::routes())
            .add_route(controllers::openapi::routes())
    }

    async fn after_context(ctx: AppContext) -> Result<AppContext> {
        create_necessary_folders()?;

        Ok(AppContext {
            cache: Cache::new(cache::drivers::inmem::new()).into(),
            storage: Storage::single(storage::drivers::local::new_with_prefix(STORAGE_FOLDER)?).into(),
            ..ctx
        })
    }

    async fn connect_workers(ctx: &AppContext, queue: &Queue) -> Result<()> {
        queue.register(DownloadWorker::build(ctx)).await?;
        Ok(())
    }
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::seed::SeedData);
        // tasks-inject (do not remove)
    }

    async fn truncate(db: &DatabaseConnection) -> Result<()> {
        // TODO add all other tables
        truncate_table(db, users::Entity).await?;
        Ok(())
    }

    async fn seed(db: &DatabaseConnection, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(db, &base.join("users.yaml").display().to_string()).await?;
        Ok(())
    }
}
