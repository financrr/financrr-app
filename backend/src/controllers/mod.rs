pub mod openapi;
pub mod session;
pub mod status;
pub mod user;
