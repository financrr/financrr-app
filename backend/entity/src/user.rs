//! `SeaORM` Entity, @generated by sea-orm-codegen 1.0.0-rc.4

use sea_orm::entity::prelude::*;
use sea_orm::ActiveValue::Set;
use serde::{Deserialize, Serialize};

use crate::error::EntityError;
use crate::utility::hashing::hash_string;
use crate::utility::time::get_now;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "user")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    #[sea_orm(column_type = "Text", unique)]
    pub username: String,
    #[sea_orm(column_type = "Text", nullable, unique)]
    pub email: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub display_name: Option<String>,
    #[sea_orm(column_type = "Text")]
    pub password: String,
    pub created_at: TimeDateTimeWithTimeZone,
    pub is_admin: bool,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::budget::Entity")]
    Budget,
    #[sea_orm(has_many = "super::currency::Entity")]
    Currency,
    #[sea_orm(has_many = "super::permissions::Entity")]
    Permissions,
    #[sea_orm(has_many = "super::session::Entity")]
    Session,
}

impl Related<super::budget::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Budget.def()
    }
}

impl Related<super::currency::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Currency.def()
    }
}

impl Related<super::permissions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Permissions.def()
    }
}

impl Related<super::session::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Session.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

impl Model {
    pub fn verify_password(&self, password: &[u8]) -> Result<bool, EntityError> {
        argon2::verify_encoded(&self.password, password).map_err(EntityError::HashingFailed)
    }
}

impl ActiveModel {
    pub fn register(
        username: String,
        email: Option<String>,
        display_name: Option<String>,
        password: String,
    ) -> Result<Self, EntityError> {
        let hashed_password = hash_string(&password)?;

        Ok(Self {
            id: Default::default(),
            username: Set(username),
            email: Set(email),
            display_name: Set(display_name),
            password: Set(hashed_password),
            created_at: Set(get_now()),
            is_admin: Set(false),
        })
    }
}

impl Entity {
    pub fn find_by_username(username: &str) -> Select<Self> {
        Self::find().filter(Column::Username.eq(username.to_string()))
    }
}
