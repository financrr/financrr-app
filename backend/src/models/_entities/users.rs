//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.3

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "users")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i64,
    #[sea_orm(column_type = "Text", unique)]
    pub email: String,
    #[sea_orm(column_type = "Text")]
    pub name: String,
    pub flags: i32,
    #[sea_orm(column_type = "Text")]
    pub password: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub reset_token: Option<String>,
    pub reset_sent_at: Option<DateTimeWithTimeZone>,
    #[sea_orm(column_type = "Text", nullable)]
    pub email_verification_token: Option<String>,
    pub email_verification_sent_at: Option<DateTimeWithTimeZone>,
    pub email_verified_at: Option<DateTimeWithTimeZone>,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::categories::Entity")]
    Categories,
    #[sea_orm(has_many = "super::currencies::Entity")]
    Currencies,
    #[sea_orm(has_many = "super::sessions::Entity")]
    Sessions,
    #[sea_orm(has_many = "super::tags::Entity")]
    Tags,
    #[sea_orm(has_many = "super::user_permissions::Entity")]
    UserPermissions,
}

impl Related<super::categories::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Categories.def()
    }
}

impl Related<super::currencies::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Currencies.def()
    }
}

impl Related<super::sessions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Sessions.def()
    }
}

impl Related<super::tags::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Tags.def()
    }
}

impl Related<super::user_permissions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::UserPermissions.def()
    }
}
