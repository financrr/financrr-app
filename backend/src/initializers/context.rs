use crate::utils::type_name::type_name_only;
use async_trait::async_trait;
use loco_rs::app::AppContext;
use loco_rs::prelude::Initializer;
use std::sync::OnceLock;

static GLOBAL_APP_CONTEXT: OnceLock<AppContext> = OnceLock::new();

/// This method should not be used if not necessary!
/// Use "try_get_global_app_context" when possible!
///
/// It was introduced to get the AppContext inside hooks like ActiveModelBehavior.
pub fn get_global_app_context() -> &'static AppContext {
    GLOBAL_APP_CONTEXT
        .get()
        .expect("GLOBAL_APP_CONTEXT should be set by now!")
}

pub fn try_get_global_app_context() -> Option<&'static AppContext> {
    GLOBAL_APP_CONTEXT.get()
}

pub struct ContextInitializer;

#[async_trait]
impl Initializer for ContextInitializer {
    fn name(&self) -> String {
        type_name_only::<Self>().to_string()
    }

    async fn before_run(&self, ctx: &AppContext) -> loco_rs::Result<()> {
        let _ = GLOBAL_APP_CONTEXT.get_or_init(|| ctx.clone());

        Ok(())
    }
}
