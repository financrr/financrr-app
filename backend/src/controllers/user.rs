use crate::error::app_error::AppError;
use crate::models::users::{Model, RegisterParams};
use crate::services::snowflake_generator::SnowflakeGenerator;
use crate::services::user_verification::UserVerificationService;
use crate::views::user::UserResponse;
use axum::extract::State;
use axum::http::StatusCode;
use axum::routing::post;
use axum::{debug_handler, Extension, Json};
use loco_rs::app::AppContext;
use loco_rs::prelude::Routes;
use sea_orm::IntoActiveModel;
use serde::Deserialize;
use utoipa::ToSchema;
use validator::{Validate, ValidationError, ValidationErrors};

#[derive(Debug, Deserialize, ToSchema)]
pub struct VerifyParams {
    pub token: String,
}

/// Registers a new User.
#[utoipa::path(post,
    path = "/api/v1/users/register",
    tag = "User",
    responses(
        (status = StatusCode::CREATED, description = "Successfully registered a new User.", content_type="application/json", body = UserResponse),
        (status = StatusCode::BAD_REQUEST, description = "Validation error.", content_type="application/json", body = AppError),
    )
)]
#[debug_handler]
async fn register(
    State(ctx): State<AppContext>,
    Extension(user_verification_service): Extension<UserVerificationService>,
    Extension(snowflake_generator): Extension<SnowflakeGenerator>,
    Json(params): Json<RegisterParams>,
) -> Result<(StatusCode, Json<UserResponse>), AppError> {
    params.validate()?;
    if !Model::is_email_unique(&ctx.db, &params.email).await? {
        let mut errors = ValidationErrors::new();
        errors.add("email", ValidationError::new("Email already exists."));

        return Err(errors.into());
    }

    let res = Model::create_with_password(&ctx.db, &snowflake_generator, &params).await?;
    let active_model = res.into_active_model();

    let model = user_verification_service
        .send_verification_email_or_verify_user(active_model)
        .await?;

    Ok((StatusCode::CREATED, Json(UserResponse::from(model))))
}

/// Verifies a User with the given token.
#[utoipa::path(post,
    path = "/api/v1/users/verify",
    tag = "User",
    responses(
        (status = StatusCode::OK, description = "Successfully verified a User.", content_type="application/json", body = UserResponse),
        (status = StatusCode::BAD_REQUEST, description = "Invalid verification token.", content_type="application/json", body = AppError),
    )
)]
#[debug_handler]
async fn verify(
    State(ctx): State<AppContext>,
    Json(params): Json<VerifyParams>,
) -> Result<(StatusCode, Json<UserResponse>), AppError> {
    let user = Model::find_by_verification_token(&ctx.db, &params.token).await?;
    match user {
        None => Err(AppError::InvalidVerificationToken()),
        Some(user) => {
            let active_model = user.into_active_model();
            let user = active_model.verified(&ctx.db).await?;

            Ok((StatusCode::OK, Json(UserResponse::from(user))))
        }
    }
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/users")
        .add("/register", post(register))
        .add("/verify", post(verify))
}
