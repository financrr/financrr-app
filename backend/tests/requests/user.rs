use crate::helpers::init::init_test;
use crate::helpers::users::{clean_up_user_model, create_user_with_email, generate_unactivated_user};
use axum::http::StatusCode;
use financrr::app::App;
use financrr::controllers::user::{ForgotParams, ResetParams, VerifyParams};
use financrr::models::users;
use financrr::utils::context::AdditionalAppContextMethods;
use financrr::views::user::UserResponse;
use insta::{assert_debug_snapshot, assert_json_snapshot, with_settings};
use loco_rs::testing;
use serde_json::json;
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("user_request");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_register() {
    init_test!();

    testing::request::<App, _, _>(|request, ctx| async move {
        assert!(ctx.is_mailer_enabled());
        const EMAIL: &str = "can.register@financrr.test";
        let payload = json!({
            "name": "TestUer",
            "email": EMAIL,
            "password": "Password123"
        });

        let response = request.post("/api/v1/users/register").json(&payload).await;
        assert_eq!(response.status_code(), StatusCode::CREATED);

        let user_response: UserResponse = response.json();

        let db_user = users::Model::find_by_email(&ctx.db, EMAIL).await.unwrap().unwrap();

        let mailer = ctx.mailer.unwrap();
        let deliveries = &mailer.deliveries();
        assert_eq!(deliveries.count, 1, "Exactly one email should be sent");

        with_settings!({
            filters => clean_up_user_model()
        }, {
            assert_debug_snapshot!(db_user);
        });

        with_settings!({
            filters => clean_up_user_model(),
        }, {
            assert_json_snapshot!(user_response);
        });
    })
    .await;
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_verify() {
    init_test!();

    testing::request::<App, _, _>(|request, ctx| async move {
        assert!(ctx.is_mailer_enabled());
        let user = generate_unactivated_user(&ctx).await;
        let token = user.email_verification_token.clone().unwrap();

        let payload = VerifyParams {
            email: user.email.clone(),
            token,
        };

        let response = request.post("/api/v1/users/verify").form(&payload).await;
        assert_eq!(response.status_code(), StatusCode::OK);

        let db_user = users::Model::find_by_email(&ctx.db, &user.email)
            .await
            .unwrap()
            .unwrap();

        assert!(db_user.email_verified_at.is_some());
        assert!(db_user.email_verification_token.is_none());
    })
    .await;
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_reset_password() {
    init_test!();

    testing::request::<App, _, _>(|request, ctx| async move {
        assert!(ctx.is_mailer_enabled());
        const EMAIL: &str = "can.reset.password@financrr.test";
        let user = create_user_with_email(&ctx, EMAIL).await;

        let payload = ForgotParams {
            email: user.email.clone(),
        };

        let response = request.post("/api/v1/users/forgot").form(&payload).await;
        assert_eq!(response.status_code(), StatusCode::OK);
        assert_eq!(ctx.mailer.unwrap().deliveries().count, 1);

        let db_user = users::Model::find_by_email(&ctx.db, &user.email)
            .await
            .unwrap()
            .unwrap();

        assert!(db_user.reset_token.is_some());

        const NEW_PASSWORD: &str = "NewPassword123";
        let payload = ResetParams {
            email: user.email.clone(),
            token: db_user.reset_token.unwrap(),
            password: NEW_PASSWORD.to_string(),
        };
        let response = request.post("/api/v1/users/reset").form(&payload).await;
        assert_eq!(response.status_code(), StatusCode::OK);
        let user_response: UserResponse = response.json();

        with_settings!({
            filters => clean_up_user_model(),
        }, {
            assert_json_snapshot!(user_response);
        });

        let db_user = users::Model::find_by_email(&ctx.db, &user.email)
            .await
            .unwrap()
            .unwrap();

        assert!(db_user.reset_token.is_none());

        assert!(db_user.verify_password(NEW_PASSWORD));
    })
    .await;
}
