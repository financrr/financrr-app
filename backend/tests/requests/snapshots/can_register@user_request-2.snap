---
source: tests/requests/user.rs
assertion_line: 55
expression: user_response
snapshot_kind: text
---
{
  "id": ID,
  "email": "can.register@financrr.test",
  "name": "TestUer",
  "flags": 2,
  "email_verification_sent_at": "DATEZ",
  "email_verified_at": null,
  "created_at": "DATEZ",
  "updated_at": "DATEZ"
}
